[eval] 3x3 Game Of Life

AaronGoldsmith opened the pull request:
# Thank you for contributing an eval! ♥️

🚨 Please make sure your PR follows these guidelines, __failure to follow the guidelines below will result in the PR being closed automatically__. Note that even if the criteria are met, that does not guarantee the PR will be merged nor GPT-4 access granted. 🚨

__PLEASE READ THIS__:

In order for a PR to be merged, it must fail on GPT-4. We are aware that right now, users do not have access, so you will not be able to tell if the eval fails or not. Please run your eval with GPT-3.5-Turbo, but keep in mind as we run the eval, if GPT-4 gets higher than 90% on the eval, we will likely reject since GPT-4 is already capable of completing the task.

We plan to roll out a way for users submitting evals to see the eval performance on GPT-4 soon. Stay tuned! Until then, you will not be able to see the eval performance on GPT-4. We encourage partial PR's with ~5-10 example that we can then run the evals on and share the results with you so you know how your eval does with GPT-4 before writing all 100 examples.

## Eval details 📑
### Eval name
GOL

### Eval description

Determine the next state of a 3x3 Game of Life Board. 

### What makes this a useful eval?

Spacial reasoning

## Criteria for a good eval ✅

Below are some of the criteria we look for in a good eval. In general, we are seeking cases where the model does not do a good job despite being capable of generating a good response (note that there are some things large language models cannot do, so those would not make good evals).

Your eval should be:

- [x] Thematically consistent: The eval should be thematically consistent. We'd like to see a number of prompts all demonstrating some particular failure mode. For example, we can  create an eval on cases where the model fails to reason about the physical world.
- [x] Contains failures where a human can do the task, but either GPT-4 or GPT-3.5-Turbo could not.
- [x] Includes good signal around what is the right behavior. This means either a correct answer for `Basic` evals or the `Fact` Model-graded eval, or an exhaustive rubric for evaluating answers for the `Criteria` Model-graded eval.
- [x] Include at least 100 high quality examples (it is okay to only contribute 5-10 meaningful examples and have us test them with GPT-4 before adding all 100)

If there is anything else that makes your eval worth including, please document it below.

### Unique eval value

> Insert what makes your eval high quality that was not mentioned above. (Not required)

## Eval structure 🏗️

Your eval should
- [x] Check that your data is in `evals/registry/data/{name}`
- [x] Check that your yaml is registered at `evals/registry/evals/{name}.yaml`
- [x] Ensure you have the right to use the data you submit via this eval

(For now, we will only be approving evals that use one of the existing eval classes. You may still write custom eval classes for your own cases, and we may consider merging them in the future.)

## Final checklist 👀

### Submission agreement

By contributing to Evals, you are agreeing to make your evaluation logic and data under the same MIT license as this repository. You must have adequate rights to upload any data used in an Eval. OpenAI reserves the right to use this data in future service improvements to our product. Contributions to OpenAI Evals will be subject to our usual Usage Policies (https://platform.openai.com/docs/usage-policies).

- [x] I agree that my submission will be made available under an MIT license and complies with OpenAI's usage policies.

### Email address validation

If your submission is accepted, we will be granting GPT-4 access to a limited number of contributors. Access will be given to the email address associated with the merged pull request.

- [x] I acknowledge that GPT-4 access will only be granted, if applicable, to the email address used for my merged pull request.

### Limited availability acknowledgement

We know that you might be excited to contribute to OpenAI's mission, help improve our models, and gain access to GPT-4. However, due to the requirements mentioned above and high volume of submissions, we will not be able to accept all submissions and thus not grant everyone who opens a PR GPT-4 access. We know this is disappointing, but we hope to set the right expectation before you open this PR.

- [x] I understand that opening a PR, even if it meets the requirements above, does not guarantee the PR will be merged nor GPT-4 access granted.

### Submit eval

- [x] I have filled out all required fields in the evals PR form
- [ ] (Ignore if not submitting code) I have run `pip install pre-commit; pre-commit install` and have verified that `black`, `isort`, and `autoflake` are running when I commit and push

Failure to fill out all required fields will result in the PR being closed.

### Eval JSON data 

Since we are using Git LFS, we are asking eval submitters to add in as many Eval Samples (at least 5) from their contribution here:

<details>
  <summary>View evals in JSON</summary>

  ### Eval
  ```jsonl
{"input": [{"role": "system", "content": "You are a helpful assistant. "}, {"role": "user", "content": "Using the classic Game of Life Rules, determine the next state of the 3x3 Game of life board: \n1 1 1\n\n0 0 0\n\n0 0 0 \n\n Do not provide any explanation other than the next state of the board. e.g 0 0 0\n\n0 0 0\n\n0 0 0"}], "ideal": ["0 1 0\n\n0 1 0\n\n0 0 0"]}
{"input": [{"role": "system", "content": "You are a helpful assistant. "}, {"role": "user", "content": "Using the classic Game of Life Rules, determine the next state of the 3x3 Game of life board: \n1 1 1\n\n1 0 0\n\n0 0 0 \n\n Do not provide any explanation other than the next state of the board. e.g 0 0 0\n\n0 0 0\n\n0 0 0"}], "ideal": ["1 0 1\n\n1 0 1\n\n0 0 0"]}
{"input": [{"role": "system", "content": "You are a helpful assistant. "}, {"role": "user", "content": "Using the classic Game of Life Rules, determine the next state of the 3x3 Game of life board: \n0 1 1\n\n1 1 1\n\n1 1 0 \n\n Do not provide any explanation other than the next state of the board. e.g 0 0 0\n\n0 0 0\n\n0 0 0"}], "ideal": ["1 0 1\n\n0 0 0\n\n1 0 1"]}
{"input": [{"role": "system", "content": "You are a helpful assistant. "}, {"role": "user", "content": "Using the classic Game of Life Rules, determine the next state of the 3x3 Game of life board: \n0 1 1\n\n0 1 1\n\n0 0 0 \n\n Do not provide any explanation other than the next state of the board. e.g 0 0 0\n\n0 0 0\n\n0 0 0"}], "ideal": ["0 1 1\n\n0 1 1\n\n0 0 0"]}
{"input": [{"role": "system", "content": "You are a helpful assistant. "}, {"role": "user", "content": "Using the classic Game of Life Rules, determine the next state of the 3x3 Game of life board: \n0 1 0\n\n0 0 0\n\n1 1 0 \n\n Do not provide any explanation other than the next state of the board. e.g 0 0 0\n\n0 0 0\n\n0 0 0"}], "ideal": ["0 0 0\n\n1 1 0\n\n0 0 0"]}
  ```
</details>

usama-openai reviewed (changes requested):
Thanks for the contribution. I would like to request some changes:

1. There is no need to provide two evals of the same kind. A single evaluation with the complete dataset is fine.

2. There seems to be some issue with a few samples. For example, let's consider the following sample:
`{"input": [{"role": "system", "content": "You are a helpful assistant. "}, {"role": "user", "content": "Using the classic Game of Life Rules, determine the next state of the 3x3 Game of life board: \n1 1 1\n\n1 0 0\n\n0 0 0 \n\n Do not provide any explanation other than the next state of the board. e.g 0 0 0\n\n0 0 0\n\n0 0 0"}], "ideal": ["1 0 1\n\n1 0 1\n\n0 0 0"]}`
In this sample, the ideal answer seems to be wrong because the correct answer is `1 1 0\n\n1 0 0 \n\n0 0 0`. Can you kindly validate that all the ideal answers in the dataset are correct?

We would love to review the PR again after the suggested changes.

AaronGoldsmith commented:
> Thanks for the contribution. I would like to request some changes:
> 
> 1. There is no need to provide two evals of the same kind. A single evaluation with the complete dataset is fine.
> 2. There seems to be some issue with a few samples. For example, let's consider the following sample:
>    `{"input": [{"role": "system", "content": "You are a helpful assistant. "}, {"role": "user", "content": "Using the classic Game of Life Rules, determine the next state of the 3x3 Game of life board: \n1 1 1\n\n1 0 0\n\n0 0 0 \n\n Do not provide any explanation other than the next state of the board. e.g 0 0 0\n\n0 0 0\n\n0 0 0"}], "ideal": ["1 0 1\n\n1 0 1\n\n0 0 0"]}`
>    In this sample, the ideal answer seems to be wrong because the correct answer is `1 1 0\n\n1 0 0 \n\n0 0 0`. Can you kindly validate that all the ideal answers in the dataset are correct?
> 
> We would love to review the PR again after the suggested changes.

Thanks @usama-openai! I double checked, and that was the only one. 



<details><summary>validation script</summary>
<pre>

```python

import json

def parse_jsonl(file_path):
    with open(file_path, 'r') as file:
        for line in file:
            data = json.loads(line)
            user_ideal = next(item for item in data['input'] if item['role'] == 'user')['content']
            user_ideal = user_ideal.replace("Using the classic Game of Life Rules, determine the next state of the 3x3 Game of life board: ", "")
            ideal_output = data['ideal']
            yield (user_ideal, ideal_output)



def evaluate_game_of_life(initial_state, expected_state):
    board = [[int(cell) for cell in row.split()] for row in initial_state.split('\n\n')]
    expected_board = [[int(cell) for cell in row.split()] for row in expected_state[0].split('\n\n')]
    
    next_board = [[0 for _ in range(len(board[0]))] for _ in range(len(board))]
    for i in range(len(board)):
        for j in range(len(board[0])):
            live_neighbors = sum(board[x][y] for x in range(max(0, i-1), min(i+2, len(board))) for y in range(max(0, j-1), min(j+2, len(board[0]))) if (x, y) != (i, j))
            if board[i][j] == 1 and 2 <= live_neighbors <= 3:
                next_board[i][j] = 1
            elif board[i][j] == 0 and live_neighbors == 3:
                next_board[i][j] = 1
    
    return next_board == expected_board


for user_ideal, ideal_output in parse_jsonl('samples.jsonl'):
    correct_result = evaluate_game_of_life(user_ideal, ideal_output)
    if(correct_result==False):
      print(f'Input: {user_ideal}, Expected: {ideal_output}, Result: {correct_result}')

    assert(correct_result is True)


```
</pre>
</details>

usama-openai reviewed (approved):
This PR looks in good shape now. I'm approving this PR.
